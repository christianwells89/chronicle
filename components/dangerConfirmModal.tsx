import {
  Button,
  Modal,
  ModalBody,
  ModalContent,
  ModalFooter,
  ModalOverlay,
  Text,
} from '@chakra-ui/react';

interface DangerConfirmModalProps {
  isOpen: boolean;
  context: string;
  consequences: string;
  onConfirm(): void;
  onClose(): void;
}

export const DangerConfirmModal: React.FC<DangerConfirmModalProps> = ({
  isOpen,
  context,
  consequences,
  onConfirm,
  onClose,
}) => {
  const handleConfirm = () => {
    onClose();
    onConfirm();
  };

  return (
    <Modal isOpen={isOpen} onClose={onClose}>
      <ModalOverlay />
      <ModalContent>
        <ModalBody>
          <Text mt={4} mb={6}>
            Are you sure you want to {context}?
          </Text>
          <Text>{consequences}</Text>
        </ModalBody>
        <ModalFooter>
          <Button variant="ghost" mr={3} onClick={onClose}>
            Cancel
          </Button>
          <Button colorScheme="red" onClick={handleConfirm}>
            Confirm
          </Button>
        </ModalFooter>
      </ModalContent>
    </Modal>
  );
};
