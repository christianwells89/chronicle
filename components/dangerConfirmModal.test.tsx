import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { DangerConfirmModal } from './dangerConfirmModal';

describe('DangerConfirmModal', () => {
  test('should not show when not open', () => {
    renderComponent({ isOpen: false });

    expect(screen.queryByText('Confirm')).not.toBeInTheDocument();
  });

  test('should include given context and consequences', () => {
    renderComponent({
      context: 'do that',
      consequences: 'I would not recommend it',
    });

    expect(screen.getByRole('dialog')).toMatchSnapshot();
  });

  test.each([['Cancel'], ['Confirm']])(
    'should close the modal when %s is clicked',
    async (label) => {
      const onClose = jest.fn();

      renderComponent({ onClose });
      await userEvent.click(screen.getByText(label));

      expect(onClose).toHaveBeenCalled();
    },
  );

  test('should not confirm when cancel is clicked', async () => {
    const onConfirm = jest.fn();

    renderComponent({ onConfirm });
    await userEvent.click(screen.getByText('Cancel'));

    expect(onConfirm).not.toHaveBeenCalled();
  });

  test('should confirm action when confirm is clicked', async () => {
    const onConfirm = jest.fn();

    renderComponent({ onConfirm });
    await userEvent.click(screen.getByText('Confirm'));

    expect(onConfirm).toHaveBeenCalled();
  });
});

type ComponentProps = React.ComponentProps<typeof DangerConfirmModal>;

function renderComponent({
  isOpen = true,
  context = 'close',
  consequences = 'This would be bad',
  onConfirm = jest.fn(),
  onClose = jest.fn(),
}: Partial<ComponentProps>) {
  render(
    <DangerConfirmModal
      isOpen={isOpen}
      context={context}
      consequences={consequences}
      onConfirm={onConfirm}
      onClose={onClose}
    />,
  );
}
