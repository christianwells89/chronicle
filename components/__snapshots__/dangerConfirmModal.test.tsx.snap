// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DangerConfirmModal should include given context and consequences 1`] = `
<section
  aria-describedby="chakra-modal--body-:r1:"
  aria-modal="true"
  class="chakra-modal__content css-6yiygt"
  id="chakra-modal-:r1:"
  role="dialog"
  style="opacity: 0; transform: scale(0.95) translateZ(0);"
  tabindex="-1"
>
  <div
    class="chakra-modal__body css-0"
    id="chakra-modal--body-:r1:"
  >
    <p
      class="chakra-text css-1puzibz"
    >
      Are you sure you want to 
      do that
      ?
    </p>
    <p
      class="chakra-text css-0"
    >
      I would not recommend it
    </p>
  </div>
  <footer
    class="chakra-modal__footer css-192qrng"
  >
    <button
      class="chakra-button css-eoelga"
      type="button"
    >
      Cancel
    </button>
    <button
      class="chakra-button css-4xx2wk"
      type="button"
    >
      Confirm
    </button>
  </footer>
</section>
`;
